### ISSUE TITLE
How to test only with custom XSS payload?

### ISSUE THREAD

**Issue by fromCharCode-git** [hahwul/dalfox#136]
Hi, good work 
I have a question, I just cant make custom XSS payloads to work.

by using 
`dalfox url https://xss-game.appspot.com/level1/frame?query= -b https://keev.es/XSS.txt`
I will get 
```
[V] Triggered XSS Payload (found DOM Object): query=<scRipT class=dalfox>confirm(45)</script>
    13 line:  s were found for <b><scRipT class=dalfox>confirm(45)</script></b>. <a href='?'>T
[POC][V][GET] https://xss-game.appspot.com/level1/frame?query=%3CscRipT+class%3Ddalfox%3Econfirm%2845%29%3C%2Fscript%3E
```
it somehow always use Dalfox XSS payloads

Any idea what I am doing wrong?
Thank you
![1](https://user-images.githubusercontent.com/24513480/97519526-aee70b00-1999-11eb-9233-ed110919cb98.PNG)

**Comment by hahwul** [comment]
Hello again,

`-b` is the blind xss option. This is an option to detect blind xss by calling script src, etc. 

The custom payload option of dalfox is as below.

```
      --custom-payload string     Add custom payloads from file
```
e.g
```
$ cat samples/sample_custompayload.txt
"><custom/onmouseover=prompt(45)>
'><iframe src=JavaScriPt:alert(45)>

$ dalfox url https://xss-game.appspot.com/level1/frame\?query\= --custom-payload samples/sample_custompayload.txt
```
One thing to know is to add custom payload to the basic logic, not just custom payload. So among the basic detection patterns, the parameters identified in the DOM Verify are subsequently not testing another payload. (Because if you have a lot of useful payloads, you'll get a lot of logs.)

**Comment by hahwul** [comment]
and If you're curious about the actual request, use the `--proxy` option to proxy with burp suite or zap and see the request for yourself. It also includes your custom payload.

**Comment by KeevDesign** [comment]
Hi, thanks for your reply
even if I use --custom-payload string I will get the same result
Page is tested by Dalfox own payload instead of mine written in XSS.txt
![4](https://user-images.githubusercontent.com/24513480/97756990-2a60cd80-1afc-11eb-9706-d319ca03cdbd.PNG)

**Comment by hahwul** [comment]
@KeevDesign 
Oh, did you install it using snapcraft? There was a dalfox patch(v2.2.1) yesterday due to a permission problem in the snapcraft version. If you installed the snapcraft version, please test it again after updating it!
(https://github.com/hahwul/dalfox/issues/134#issuecomment-719657706)

```
$ sudo snap refresh dalfox
```

And even if you use custom-payload, the result of custom-payload may not appear because the basic payload also works. (If detected first in the basic payload)

**Comment by KeevDesign** [comment]
Hi, 
I re-installed the Parrot from 0 to be sure I have no issue with my system.
I installed the dalfox by snap install dalfox --devmode and checked by sudo snap refresh dalfox to be sure I have the last version.
I have 2 issues
1. I still cant test with only my payload, dalfox payload is always tested first. Is there a way how to test only my XSS payload without using the one used by dalfox?
I used this command
![1](https://user-images.githubusercontent.com/24513480/97793293-80ac3a00-1bea-11eb-89fd-a0d1cc8a9d0c.PNG)

2. when I try to use list of URLs it gives me an error " unsupported protocol scheme"
my command:
dalfox file urls --custom-payload XSS.txt
![2](https://user-images.githubusercontent.com/24513480/97793333-e7315800-1bea-11eb-802c-61e75d630685.PNG)
any idea?
thank you for your work and time

**Comment by hahwul** [comment]
@KeevDesign 
Hello again
Oh, I misunderstood the intention of the question a little bit. I'm sorry for a bit of a deadbeat response.

1) Currently, we do not have a function to test custom payload only. It seems like a necessary function, so I'll add it as an issue!

2) The result of reading the file from the message `unsupported protocol scheme ""` seems to be blank. I think, snapcraft file read permission(https://github.com/hahwul/dalfox/issues/134).. Could you check the version information?

```
1) dalfox version
$ dalfox version

2) dalfox snapcraft permission
$ snap connections | grep dalfox 
home                      dalfox:home                               :home                            -
network                   dalfox:network                            :network                         -
network-bind              dalfox:network-bind                       :network-bind                    -
```

**Comment by KeevDesign** [comment]
Hi @hahwul 

1. excellent, I think custom payload testing will be interesting function
2. Target URL: seems to be name of the file and not the content of the file
the same behavior as reported by
https://github.com/hahwul/dalfox/issues/134#

here my version and permissions of dalfox
![1](https://user-images.githubusercontent.com/24513480/97814954-2a430800-1c8b-11eb-956a-e5645b5f9d58.PNG)

best regards

**Comment by hahwul** [comment]
@KeevDesign 
That's strange. I'll test it by create the environment similarly!

**Comment by hahwul** [comment]
@KeevDesign 
Hello again,
I thought about it when I saw what we talked about, but is there a url list in the home directory? (I think it's /opt on the screen shot) Snapcraft is managing the perm for the directory, so dalfox has rights only to the bottom of the HOME directory through the `home interface`. That's the bottom part. 

```
home                      dalfox:home                               :home                            -
```
https://snapcraft.io/docs/home-interface

Will there be an error even if I test the URL list after placing it in the lower part of the home directory?

**Comment by hahwul** [comment]
@KeevDesign 
If you think you need to use it outside of the home directory, I think installing `go-get` would be much more useful than snapcraft.

```
$ go get -u github.com/hahwul/dalfox
$ ~/go/bin/dalfox
```

**Comment by KeevDesign** [comment]
thanks hahwul, you are right the snap permissions were the issue here. I am running now from Home directory and its recognizing the custom payload file and its content.
Can I ask you one more question please 
The program recognize my custom XSS file (and number of lines inside) but looks like its not testing with them.
`Added your 42 custom xss payload
[*] Start XSS Scanning.. with 0 queries 🗡
[*] Finish :D
`
Do I need to specify the parameter point for such XSS testing?
![1](https://user-images.githubusercontent.com/24513480/98044945-02d57200-1e28-11eb-9b75-e372d84b847e.PNG)
 
thank you

**Comment by spook95** [comment]
☝🏾🌎💚

Sent from my iPhone

On 3 Nov 2020, at 21:59, KeevDesign <notifications@github.com> wrote:

﻿

thanks hahwul, you are right the snap permissions were the issue here. I am running now from Home directory and its recognizing the custom payload file and its content.
Can I ask you one more question please
The program recognize my custom XSS file (and number of lines inside) but looks like its not testing with them.
Added your 42 custom xss payload [*] Start XSS Scanning.. with 0 queries 🗡 [*] Finish :D
Do I need to specify the parameter point for such XSS testing?
[1]<https://user-images.githubusercontent.com/24513480/98044945-02d57200-1e28-11eb-9b75-e372d84b847e.PNG>

thank you

**Comment by spook95** [comment]
Sorry I don’t know if I done it ?

**Comment by hahwul** [comment]
@fromCharCode-git 
Well, I think I found a similar situation. I'm guess. it didn't scan the parameters found with Mining. I am on it.
Thank you always for the issue report!

with parameter
```
$ dalfox url https://xss-game.appspot.com/level1/frame\?query\=j --only-custom-payload --custom-payload 123
...
[*] Added your 1 custom xss payload
[*] Start XSS Scanning.. with 3 queries 🗡
[*] Finish :D
```

without parameter (found query param from mining)
```
$ dalfox url https://xss-game.appspot.com/level1/frame --only-custom-payload --custom-payload 123
[*] Added your 1 custom xss payload
[*] Start XSS Scanning.. with 0 queries 🗡
[*] Finish :D
```

**Comment by hahwul** [comment]
@fromCharCode-git 
I patched that issue! Update to v2.2.3 and test it!

```
$ go get -u github.com/hahwul/dalfox
```

```
$ sudo snap refresh dalfox
```

```
$ brew upgrade dalfox
```
